File: esw-ocs/esw-ocs-dsl/src/test/scala/esw/dsl/script/EventServiceDslTest.scala
package esw.dsl.script

import csw.event.api.scaladsl.{EventPublisher, EventService, EventSubscriber}
import csw.params.core.models.Prefix
import csw.params.events.{Event, EventKey, EventName, SystemEvent}
import esw.dsl.script.services.EventServiceDsl
import esw.ocs.api.BaseTestSuite
import esw.ocs.macros.StrandEc
import org.mockito.ArgumentMatchers.{any, eq => argsEq}

import scala.concurrent.Future
import scala.concurrent.duration.DurationDouble

class EventServiceDslTest extends BaseTestSuite with EventServiceDsl {
  val eventService: EventService  = mock[EventService]
  private val eventPublisher      = mock[EventPublisher]
  private val eventSubscriber     = mock[EventSubscriber]
  private val event               = SystemEvent(Prefix("TCS.test"), EventName("event-1"))
  implicit val strandEc: StrandEc = StrandEc()

  when(eventService.defaultPublisher).thenReturn(eventPublisher)
  when(eventService.defaultSubscriber).thenReturn(eventSubscriber)

  "publish" must {
    "delegate to publishing single event | ESW-120" in {
      publishEvent(event)
      verify(eventService.defaultPublisher).publish(event)
    }

    "delegate to publishing event with generator | ESW-120" in {
      publishEvent(5.seconds)(Some(event))
      verify(eventService.defaultPublisher).publishAsync(
        any[Future[Option[Event]]],
        argsEq(5.seconds)
      )
    }
  }

  "subscribe" must {
    "delegate to subscribing events | ESW-120" in {
      onEvent("TCS.test.event-1")(println)
      verify(eventService.defaultSubscriber).subscribeAsync(any[Set[EventKey]], any[Event => Future[_]]())
    }
  }

  "get" must {
    "delegate to getting events | ESW-120" in {
      getEvent("TCS.test.event-1")
      verify(eventService.defaultSubscriber).get(any[Set[EventKey]])
    }
  }
}

File: esw-kt/script-dsl/src/test/kotlin/esw/ocs/dsl/highlevel/EventServiceDslTest.kt
package esw.ocs.dsl.highlevel

import akka.Done.done
import akka.actor.Cancellable
import csw.event.api.javadsl.IEventPublisher
import csw.event.api.javadsl.IEventSubscriber
import csw.event.api.javadsl.IEventSubscription
import csw.params.core.models.Prefix
import csw.params.events.*
import io.kotlintest.shouldBe
import io.kotlintest.specs.WordSpec
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import java.util.concurrent.CompletableFuture
import kotlin.coroutines.EmptyCoroutineContext
import kotlin.time.milliseconds
import kotlinx.coroutines.CoroutineScope

class EventServiceDslTest : WordSpec(), EventServiceDsl {

    private val key = "TCS.test.eventkey1"
    private val duration = 10.milliseconds

    private val event: Event = mockk<ObserveEvent>()
    private val eventKeys = HashSet<EventKey>()
    private val eventSet = HashSet<Event>(1)

    private val eventSubscription = mockk<IEventSubscription>()
    private val cancellable = mockk<Cancellable>()
    private val eventCallback = mockk<(Event) -> CompletableFuture<*>>()
    private val eventPublisher: IEventPublisher = mockk()
    private val eventSubscriber: IEventSubscriber = mockk()

    override val coroutineScope: CoroutineScope = CoroutineScope(EmptyCoroutineContext)
    override val defaultPublisher: IEventPublisher = eventPublisher
    override val defaultSubscriber: IEventSubscriber = eventSubscriber

    init {
        eventKeys.add(EventKey.apply(key))
        eventSet.add(event)
        "EventServiceDsl" should {

            "systemEvent should return a SystemEvent created with given parameters | ESW-120" {
                val eventName = "systemEvent1"
                val eventPrefix = "TCS.filter.wheel"
                val systemEvent = systemEvent(eventPrefix, eventName)

                // Verify that  event with provided prefix and eventName is created.
                systemEvent shouldBe SystemEvent(
                    systemEvent.eventId(),
                    Prefix(eventPrefix),
                    EventName(eventName),
                    systemEvent.eventTime(),
                    systemEvent.paramSet()
                )
            }

            "observeEvent should return a ObserveEvent created with given parameters | ESW-120" {
                val eventName = "observeEvent1"
                val eventPrefix = "TCS.filter.wheel"
                val observeEvent = observeEvent(eventPrefix, eventName)

                // Verify that event with provided prefix and eventName is created.
                observeEvent shouldBe ObserveEvent(
                    observeEvent.eventId(),
                    Prefix(eventPrefix),
                    EventName(eventName),
                    observeEvent.eventTime(),
                    observeEvent.paramSet()
                )
            }

            "publish should delegate to publisher.publish | ESW-120" {
                every { (eventPublisher.publish(event)) }.returns(CompletableFuture.completedFuture(done()))
                publishEvent(event) shouldBe done()
                verify { eventPublisher.publish(event) }
            }

            "publishEvent should delegate to publisher.publishAsync | ESW-120" {
                every { eventPublisher.publishAsync(any(), any()) }.answers { cancellable }
                publishEvent(duration) { event } shouldBe cancellable
                verify { eventPublisher.publishAsync(any(), any()) }
            }

            "onEvent should delegate to subscriber.subscribeAsync | ESW-120" {
                every { eventSubscriber.subscribeAsync(eventKeys, any()) }.answers { eventSubscription }
                onEvent(key) { eventCallback } shouldBe eventSubscription
                verify { eventSubscriber.subscribeAsync(eventKeys, any()) }
            }

            "getEvent should delegate to subscriber.get | ESW-120" {
                every { eventSubscriber.get(eventKeys) }.answers { CompletableFuture.completedFuture(eventSet) }
                getEvent(key) shouldBe eventSet
                verify { eventSubscriber.get(eventKeys) }
            }
        }
    }
}
